crates/bevy_ecs/src/reflect/mod.rs

```diff
@@ -45,12 +46,13 @@ impl DerefMut for AppTypeRegistry {
 /// Creates a `T` from a `&dyn Reflect`.
 ///
 /// The first approach uses `T`'s implementation of `FromReflect`.
-/// If this fails, it falls back to default-initializing a new instance of `T` using its
-/// `ReflectFromWorld` data from the `world`'s `AppTypeRegistry` and `apply`ing the
+/// If this fails, it falls back to default-initializing a new instance of `T` using
+/// either its `ReflectDefault` or its `ReflectFromWorld` registration in `registry`
+/// (whichever is found, preferring `ReflectDefault` if present) and `apply`ing the
 /// `&dyn Reflect` on it.
 ///
-/// Panics if both approaches fail.
-fn from_reflect_or_world<T: FromReflect>(
+/// Panics if all approaches fail.
+fn from_reflect_with_fallback<T: FromReflect>(
```

Comment:
Since this essentially means a `ReflectComponent` type can rely on any of `FromReflect`, `Default`, or `FromWorld`, I wonder if we should remove the `FromReflect` bound and replace it with the dynamic `ReflectFromReflect` type data.

Most users deriving `Reflect` wouldn't notice a difference. And those who know how to opt-out of `FromReflect` (with `#[reflect(from_reflect = false)]`) can still benefit from `ReflectComponent` and friends.

Reply:
Yeah that seems a good idea. The only downside I can think of is if the user passes a `TypeRegistry` in which `T` is not registered in (previously this would have worked if the `FromReflect` `impl` didn't fail).

Reply:
Yeah, I think it should be on the user to provide a `TypeRegistry` with all the necessary types.

You don't have to add this to this PR if you'd prefer it be a separate thing. I'm okay with holding off if we want to get other discussion going.

Reply:
I've changed them to use the dynamic `ReflectFromReflect`, though one downside I've now noticed is that `ReflectBundle::insert` now needs an additional `&TypeRegistry` parameter which was missed in the previous PR (my bad)